### ISSUE TITLE
Numerical mismatch in torch.sum(dim=1) on CPU between 0.3.1 and 0.4.1

### ISSUE THREAD

**Issue by volcacius** [pytorch/pytorch#10682]
## Issue description

Torch.sum(x, dim=1) with x a 2d tensor gives a different result on pytorch 0.4.0/1 compared to pytorch 0.3.1. I couldn't find any documentation/issue on this change of behavior, so I'm opening this issue to track it. Any feedback or links to related issues that I may have missed would help.
It's a small difference, but it appears to be the reason why I loose 0.02 top-1 accuracy on a CIFAR10 model simply by upgrading pytorch.

#### Pytorch 0.3.1: 

In [1]: import torch

In [2]: torch.manual_seed(1)
Out[2]: <torch._C.Generator at 0x7f75ed7fde30>

In [3]: torch.set_printoptions(precision=12)

In [4]: r = torch.rand(18,18)

In [5]: s = torch.sum(r, dim=1)

In [6]: print(s[0])
9.30374813079834

#### Pytorch 0.4.1:

 In [1]: import torch

In [2]: torch.manual_seed(1)
Out[2]: <torch._C.Generator at 0x7fb075b08b50>

In [3]: torch.set_printoptions(precision=12)

In [4]: r = torch.rand(18,18)

In [5]: s = torch.sum(r, dim=1)

In [6]: print(s[0])
tensor(9.303749084473)

#### Pytorch 0.3.1
Running in a docker image built from the official Dockerfile for tag v0.3.1

PyTorch version: 0.3.1b0+2b47480
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration: GPU 0: Tesla P4
Nvidia driver version: 387.26
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21
/usr/lib/x86_64-linux-gnu/libcudnn_static_v6.a

Versions of relevant libraries:
[pip] numpy (1.15.0)
[pip] torch (0.3.1b0+2b47480)
[pip] torchvision (0.2.1)
[conda] Could not collect

#### Pytorch 0.4.1
Running in the official pytorch Docker image (pytorch/pytorch:0.4.1-cuda9-cudnn7-runtime, 7b329a33d981)

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Tesla P4
Nvidia driver version: 387.26
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy (1.14.5)
[pip] torch (0.4.1)
[pip] torchvision (0.2.1)
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] pytorch                   0.4.1           py36_cuda9.0.176_cudnn7.1.2_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by zou3519** [comment]
This makes sense because of floating point accuracy (to 1e-5 or 1e-6, don't remember which one). If you want it to have higher precision, you can use the double type.
